### DISCUSSION TITLE
Using torchio for DDP

### DISCUSSION THREAD

**Discussion by alabamagan** [TorchIO-project/torchio#1012] [Q&A]
I can't seem to find any discussion of using torchio with distributed model. I think the challenge mainly lies at how the dataloader samplers can distribute between processes of different ranks. Does anyone has the idea how this can be achieved with torchio?

**Comment by fepegar** [comment]
I think @hsyang1222 might be able to help. See also
- #1021

**Reply by alabamagan** [reply]
Fabulous, I currently just create separate `Queue` on separate process and you can imagine how messy it would get. I will have a look at the PR.

**Comment by hsyang1222** [comment]
We recently added feature for support DDP samplers in torchio queues. Please see the queue section of the updated documentation. If you don't understand the documentation, please let me know.

**Comment by alabamagan** [comment]
Seems like the `Queue` in each process will loop through the whole data set so that if you are DDP training with a world size N, each epoch will be trained with N x [num_samples]. Is this the default behavior or am I missing something?

**Reply by hsyang1222** [reply]
As the world size _N_ increases, the number of samples learned in each epoch _num_samples_ remains the same.

First, without using DDP, _num_samples_ would be calculated as follows.

Number of subjects _S_ * number of patch extractions per subject (samples_per_volume) _P_ = total number of samples _num_samples_

When using a distributed sampler and queue in an environment with a DDP of world size _N_, the distributed sampler distributes the subjects to be used at each rank. In each rank, _num_samples_ is calculated as follows

_num_samples_ in rank 0 : (_S_ // _N_) * _P_ 
_num_samples_ in rank 1 : (_S_ // _N_) * _P_
 ...
_num_samples_ in rank N-1 : (_S_ // _N_) * _P_

The sum of these is (_S_ // _N_) * _P_ * _N_, which is the same number as without DDP. Of course, there is some margine when _S_ is not divisible by _N_, but it is negligible.
